frame/elections/src/lib.rs

```diff
@@ -890,11 +914,77 @@ impl<T: Config> Pallet<T> {
 		debug_assert!(_remainder.is_zero());
 	}
 
-	/// Run the phragmen election with all required side processes and state updates, if election
-	/// succeeds. Else, it will emit an `ElectionError` event.
+	/// Run an election with all required side processes and state updates, if election
+	/// succeeds. Else, it will emit an `ElectionError` event. The election algorithm is defined
+	/// by the implementor of `Self::ElectionSolver`.
 	///
 	/// Calls the appropriate [`ChangeMembers`] function variant internally.
-	fn do_phragmen() -> Weight {
+	fn do_election() -> Weight {
+		let PreElectionResults {
+			num_to_elect,
+			candidate_ids,
+			candidates_and_deposit,
+			voters_and_stakes,
+			voters_and_votes,
+			num_edges,
+		} = match Self::do_pre_solve_election() {
+			Ok(results) => results,
+			Err(err) => match err {
+				Error::EmptyTerm => {
+					Self::deposit_event(Event::EmptyTerm);
+					return T::DbWeight::get().reads(3)
+				},
+				Error::TooManyVotes => {
+					Self::deposit_event(Event::ElectionError);
```

Comment:
May be we can emit a different event here, `Event::PreElectionError`.

Reply:
The concept of pre/post election is only relevant for weight calculation and it should be transparent to users. I think that exposing a `Event::PreElectionError` event would propagate the pre/post election concept to the user, which probably we don't want to to.